libdnf/rpm/repo_sack.cpp

```diff
@@ -19,22 +19,58 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 
 #include "libdnf/rpm/repo_sack.hpp"
 
+#include "repo_impl.hpp"
+
 #include "libdnf/base/base.hpp"
 #include "libdnf/conf/config_parser.hpp"
 #include "libdnf/conf/option_bool.hpp"
+#include "libdnf/rpm/solv_sack_impl.hpp"
 
 #include <fmt/format.h>
 
+extern "C" {
+#include <solv/repo.h>
+#include <solv/solv_xfopen.h>
+#include <solv/testcase.h>
+}
+
 #include <filesystem>
 
+
+using LibsolvRepo = ::Repo;
+
+
 namespace libdnf::rpm {
 
+
 RepoWeakPtr RepoSack::new_repo(const std::string & id) {
     // TODO(jrohel): Test repo exists
     auto repo = std::make_unique<Repo>(id, *base, Repo::Type::AVAILABLE);
     return add_item_with_return(std::move(repo));
 }
 
+// Deleter for std::unique_ptr<LibsolvRepo>
+static void libsolv_repo_free(LibsolvRepo * libsolv_repo) {
+    repo_free(libsolv_repo, 1);
+}
+
+
+RepoWeakPtr RepoSack::new_repo_from_libsolv_testcase(const std::string & repoid, const std::string & path) {
+    std::unique_ptr<std::FILE, decltype(&std::fclose)> testcase_file(solv_xfopen(path.c_str(), "r"), &std::fclose);
+    if (!testcase_file) {
+        throw std::runtime_error(fmt::format("Unable to open libsolv testcase file: {}", path));
```

Comment:
Could you use a proper exception class using the exception system that we have?

Reply:
replaced with `throw SystemError(EIO, "Unable to open libsolv testcase file", path)`